### ISSUE TITLE
LGTM.com - false positive C Function call implicitly declares 'strlen'.

### ISSUE THREAD

**Issue by rossphilipson** [github/codeql#9470]
In three separate warning alerts, the calls are said to be an implicit declaration, specifically calls to strlen() and memset(). The header file for the definition of strlen() is included. A colleague suspects the issue is out redirecting the definition of of strlen() to use compiler build-in functions leading to confusion. The definitions of strlen() e,g, looks like this:

`size_t strlen(const char *s);
#define strlen(s)       __builtin_strlen(s)`

https://lgtm.com/projects/g/TrenchBoot/secure-kernel-loader/snapshot/98ef0cdfa774f588c95c52162c45ce8e96df27d4/files/event_log.c?sort=name&dir=ASC&mode=heatmap#xaf0d6774fffddb7:2

**Comment by hmakholm** [comment]
You're right that this looks odd indeed. The fact that the alert complains about `strlen` rather than `__builtin_strlen` seems to indicate that it doesn't see the macro definition at all.

I suspect the underlying cause has to be that the CodeQL database extractor fails to understand that `-ffreestanding` in the compiler invocation means that `__STDC_HOSTED__` should _not_ be defined when preprocessing the source. In that case https://github.com/TrenchBoot/secure-kernel-loader/blob/bf81163/include/string.h will attempt to include _itself_ recursively, and the recursive inclusion produces nothing because its inclusion guard `__STRINGS_H__` is already defined.

@github/codeql-c-extractor, does this sound reasonable?

**Comment by jketema** [comment]
Thanks again for reporting this problem. The issue should be fixed in the next version of CodeQL which will be released in a few weeks. Closing this issue for now.
